include/handler.h

```diff
@@ -434,6 +465,27 @@ class handler {
 
 	void extend_lifetime(std::shared_ptr<detail::lifetime_extending_state> state) { m_attached_state.emplace_back(std::move(state)); }
 
+	template <int Dims>
+	void experimental_constrain_split(const range<Dims>& constraint) {
+		assert(m_task == nullptr);
+		m_split_constraint = detail::range_cast<3>(constraint);
+	}
+
+	template <int Dims>
+	range<3> get_constrained_granularity(const range<Dims>& global_size, const range<Dims>& granularity) const {
+		range<3> result = detail::range_cast<3>(granularity);
+		for(int i = 0; i < Dims; ++i) {
+			const auto lcm = std::lcm(granularity[i], m_split_constraint[i]);
+			if(lcm == 0) { throw std::runtime_error("Split constraint cannot be 0"); }
+			result[i] = lcm;
+		}
+		if(global_size % detail::range_cast<Dims>(result) != range<Dims>(detail::zeros)) {
+			throw std::runtime_error(
+			    fmt::format("The split constraint {} does not evenly divide the global size {}", detail::range_cast<Dims>(result), global_size));
```

Comment:
I would make this a bit more descriptive to the user (who might not know where the exception comes from). e.g. `The effective split constraint {} does not evenly divide the kernel global size {}`.

Reply:
Yes, good point. I've adjusted the message to only include "effective" if it's an nd-range kernel though!

Reply:
I like it!